File: bindings/python/python/wcl/_convert.py
"""JSON-to-Python value conversion for WCL WASM binding."""

import json


def json_to_py(val):
    """Convert a parsed JSON value to a Python native value."""
    if val is None:
        return None
    if isinstance(val, bool):
        return val
    if isinstance(val, int):
        return val
    if isinstance(val, float):
        # Preserve ints that JSON parsed as float
        if val == int(val) and not isinstance(val, bool):
            return int(val)
        return val
    if isinstance(val, str):
        return val
    if isinstance(val, list):
        return [json_to_py(v) for v in val]
    if isinstance(val, dict):
        # Check for set encoding
        if val.get("__type") == "set" and "items" in val:
            items = [json_to_py(v) for v in val["items"]]
            try:
                return set(items)
            except TypeError:
                return items
        # Check for block ref encoding
        if "kind" in val and ("attributes" in val or "children" in val or "decorators" in val):
            return _json_to_block_ref(val)
        return {k: json_to_py(v) for k, v in val.items()}
    return val


def json_to_values(json_str):
    """Parse a JSON string and convert to a Python dict of values."""
    data = json.loads(json_str)
    return {k: json_to_py(v) for k, v in data.items()}


def json_to_blocks(json_str):
    """Parse a JSON array string and convert to BlockRef objects."""
    from wcl._types import BlockRef
    data = json.loads(json_str)
    return [_json_to_block_ref(b) for b in data]


def json_to_diagnostics(json_str):
    """Parse a JSON array string and convert to Diagnostic objects."""
    from wcl._types import Diagnostic
    data = json.loads(json_str)
    return [
        Diagnostic(
            severity=d["severity"],
            message=d["message"],
            code=d.get("code"),
        )
        for d in data
    ]


def _json_to_block_ref(obj):
    """Convert a JSON block ref object to a BlockRef."""
    from wcl._types import BlockRef, Decorator
    attrs_raw = obj.get("attributes", {})
    attributes = {k: json_to_py(v) for k, v in attrs_raw.items()}
    children_raw = obj.get("children", [])
    children = [_json_to_block_ref(c) for c in children_raw]
    decorators_raw = obj.get("decorators", [])
    decorators = [
        Decorator(
            name=d["name"],
            args={k: json_to_py(v) for k, v in d.get("args", {}).items()},
        )
        for d in decorators_raw
    ]
    return BlockRef(
        kind=obj["kind"],
        id=obj.get("id"),
        labels=obj.get("labels", []),
        attributes=attributes,
        children=children,
        decorators=decorators,
    )

File: bindings/ruby/lib/wcl/convert.rb
require "json"

module Wcl
  # JSON-to-Ruby value conversion for WCL WASM binding.
  module Convert
    module_function

    def json_to_ruby(val)
      case val
      when nil, true, false, String
        val
      when Integer
        val
      when Float
        val == val.to_i && !val.is_a?(Float) ? val.to_i : val
      when Array
        val.map { |v| json_to_ruby(v) }
      when Hash
        # Check for set encoding
        if val["__type"] == "set" && val.key?("items")
          items = val["items"].map { |v| json_to_ruby(v) }
          begin
            Set.new(items)
          rescue
            items
          end
        # Check for block ref encoding
        elsif val.key?("kind") && (val.key?("attributes") || val.key?("children") || val.key?("decorators"))
          json_to_block_ref(val)
        else
          val.transform_values { |v| json_to_ruby(v) }
        end
      else
        val
      end
    end

    def json_to_values(json_str)
      data = JSON.parse(json_str)
      data.transform_values { |v| json_to_ruby(v) }
    end

    def json_to_blocks(json_str)
      data = JSON.parse(json_str)
      data.map { |b| json_to_block_ref(b) }
    end

    def json_to_diagnostics(json_str)
      data = JSON.parse(json_str)
      data.map do |d|
        Diagnostic.new(
          severity: d["severity"],
          message: d["message"],
          code: d["code"]
        )
      end
    end

    def json_to_block_ref(obj)
      attrs = (obj["attributes"] || {}).transform_values { |v| json_to_ruby(v) }
      children = (obj["children"] || []).map { |c| json_to_block_ref(c) }
      decorators = (obj["decorators"] || []).map do |d|
        Decorator.new(
          name: d["name"],
          args: (d["args"] || {}).transform_values { |v| json_to_ruby(v) }
        )
      end

      BlockRef.new(
        kind: obj["kind"],
        id: obj["id"],
        attributes: attrs,
        children: children,
        decorators: decorators
      )
    end
  end
end
